src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -423,6 +423,32 @@ def flatten_datasets(self, datasets):
 
     @filterable
     def query(self, filters=None, options=None):
+        """
+        In `query-options` we can provide `extra` arguments which control which data should be retrieved
+        for a dataset.
+
+        `query-options.extra.mountpoint` makes sure mountpoint property is retrieved and added to the top level
+        dictionary as a key.
+
+        `query-options.extra.properties` is a list of properties which should be retrieved. If null ( by default ),
+        it would retrieve all properties, if empty, it will retrieve no property ( `mountpoint` is special in this
+        case and is controlled by `query-options.extra.mountpoint` attribute ).
+
+        We provide 2 ways how zfs.dataset.query returns dataset's data. First is a flat structure ( default ), which
+        means that all the datasets in the system are returned as separate objects which also contain all the data
+        their is for their children. This retrieval type is slightly slower because of duplicates which exist in
+        each object.
+        Second type is hierarchical where only top level datasets are returned in the list and they contain all the
+        children there are for them in `children` key. This retrieval type is slightly faster.
+        These options are controlled by `query-options.extra.flat` attribute which defaults to true.
+
+        `query-options.extra.custom_properties` controls if user defined properties of datasets should be retrieved
```

Comment:
If uts user properties, why call it custom_properties and not user_properties?

Reply:
I was reflecting on that when i wrote the docs. I'll update this :)